provider/sma/device.go

```diff
@@ -15,58 +14,48 @@ type Device struct {
 	*sunny.Device
 
 	log    *util.Logger
-	mux    sync.Mutex
-	wait   *util.Waiter
-	values map[sunny.ValueID]interface{}
+	values *util.Monitor[map[sunny.ValueID]any]
 	once   sync.Once
 }
 
-// StartUpdateLoop if not already started
-func (d *Device) StartUpdateLoop() {
-	d.once.Do(func() {
-		go func() {
-			for range time.Tick(time.Second * 5) {
-				if err := d.UpdateValues(); err != nil {
-					d.log.ERROR.Println(err)
-				}
-			}
-		}()
-	})
+// Run starts the receive loop once per device
+func (d *Device) Run() {
+	d.once.Do(d.run)
 }
 
-func (d *Device) UpdateValues() error {
-	d.mux.Lock()
-	defer d.mux.Unlock()
+func (d *Device) run() {
+	for range time.Tick(5 * time.Second) {
+		if err := d.UpdateValues(); err != nil {
+			d.log.ERROR.Println(err)
+		}
+	}
+}
 
-	values, err := d.Device.GetValues()
+func (d *Device) UpdateValues() error {
+	res, err := d.Device.GetValues()
 	if err == nil {
-		err = mergo.Merge(&d.values, values, mergo.WithOverride)
-		d.wait.Update()
+		current, _ := d.values.Get()
+		if current == nil {
+			current = res
+		} else {
+			maps.Copy(current, res)
+		}
+		d.values.Set(current)
```

Comment:
I think this could cause a race condition if at the same time the return value of `Values()` is handled.
Based on my knowledge maps are passed by references so both goroutines  would work on the same object.